### PR TITLE
Sizing:System and AirloopHVAC objects in different order causes differences

### LINKED ISSUE #9711: Order of Sizing:System with different capacity control methods and AirloopHVAC objects causes differeneces in results
Issue overview
--------------
A large user file shows a different answer when the order of the objects is changed. The underlying cause was differences in capacity control method in the Sizing:System objects. When the index for AirloopHVAC and Sizing:System objects does not line up, the GetCoilDesFlowT looks up the wrong capacity control method and the answer changes.

There are several things that must occur for differences to occur:

1) The order of AirloopHVAC must be different from the corresponding Sizing:System objects.
2) The Sizing:System objects must use different capacity control methods and/or Type of Load to Size On.
3) The supply air conditions must be different among the various AirloopHVAC objects (Sizing:System objects).

A few lines of code corrects the problem. The new lookup to find the correct Sizing:System object and the use of the local int in 3 places (the 3rd is a warning).

    void GetCoilDesFlowT()

        int sysSizIndex = UtilityRoutines::FindItemInList(FinalSysSizing(SysNum).AirPriLoopName, state.dataSize->SysSizInput, &SystemSizingInputData::AirPriLoopName);
        if (sysSizIndex == 0) sysSizIndex = 1;

        CoolCapCtrl = SysSizInput(sysSizIndex).CoolCapControl;
        PeakLoadType = SysSizInput(sysSizIndex).CoolingPeakLoadType;

Issue and correction provided by Trane North America. These are V9.5 input files.

[Baseline_NREL_EP950.idf.txt](https://github.com/NREL/EnergyPlus/files/9805713/Baseline_NREL_EP950.idf.txt)

[Run01_NREL_EP950.idf.txt](https://github.com/NREL/EnergyPlus/files/9872451/Run01_NREL_EP950.idf.txt)

Updated defect files to V22.2:

[Baseline_NREL_EP22-2.idf.txt](https://github.com/NREL/EnergyPlus/files/9873349/Baseline_NREL_EP22-2.idf.txt)

[Run01_NREL_EP22-2.idf.txt](https://github.com/NREL/EnergyPlus/files/9873352/Run01_NREL_EP22-2.idf.txt)

Work around: Place Sizing:System and AirloopHVAC objects in the same order in the input file.

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (issue description)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by rraustad**
Pull request overview
---------------------
 - Fixes #9711
 - When Sizing:System and AirloopHVAC objects are not in the same order differences occur if Sizing:System objects have different inputs for capacity control and supply air conditions.
 - This change will also correct an undeterminant result (i.e., array access violation) when only 1 Sizing:System object is used for multiple AirloopHVAC objects where the previous code would try to access a System:Sizing object that was not present.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/DataSizing.cc
Comment (rraustad) [inline src/EnergyPlus/DataSizing.cc:459]: Basic correction to issue. SysSizInput and FinalSysSizing arrays will have different index when the Sizing:System and AirloopHVAC objects are not in the same order in idf.

**Comment by rraustad** [comment]
Comparison of defect files of Baseline and Run01 input files (Baltimore Blt Washngtn IntL,MD,USA,TMY3,724060):
Baseline defect file is left column in each comparison and this specific result from each branch is compared at the right of figure just to be sure they are the same in both cases.

image

File: src/EnergyPlus/DXCoils.cc
Comment (rraustad) [inline src/EnergyPlus/DXCoils.cc]: Debug build showed unit test failed with this variable uninitialized. Maybe my debug switches are different than linux, but initializing here corrected this unit test.

File: tst/EnergyPlus/unit/UnitarySystem.unit.cc
Comment (rraustad) [inline tst/EnergyPlus/unit/UnitarySystem.unit.cc:278]: New/corrected references in function GetCoilDesFlowT now requires the allocation of CalcSysSizing in unit test fixtures.

File: src/EnergyPlus/DataSizing.cc
Comment (rraustad) [inline src/EnergyPlus/DataSizing.cc:502]: Overall changes here are of course the issue correction (2 lines, 456 and 458) and then clean up includes getting rid of locals that could be replaced by simple referenced memory locations and using locals inside specific part of IF blocks instead of setting all the time.

File: src/EnergyPlus/DataSizing.cc
Comment (rraustad) [inline src/EnergyPlus/DataSizing.cc]: I just noticed that this IF block isn't needed since it does the same thing regardless of CoolingPeakLoadType (same for IF block at line 496). Also noticed that Latent and VentilationRequirement are not included here and not sure what was happening for those use cases.

    Sizing:System,
    A2, \field Type of Load to Size On
      \note Specifies the basis for sizing the system supply air flow rate
      \note Sensible and Total use the zone design air flow rates to size the system supply air flow rate
      \note The cooling coil will then be sized at either the peak Sensible or Total flow rate and conditions
      \note The heating coil is always sized at the peak sensible heating load.
      \note Latent uses the larger of the zone sensible and latent design air
      \note flow rates for system sizing.
      \note VentilationRequirement uses the system ventilation requirement
      \type choice
      \key Sensible
      \key Latent
      \key Total
      \key VentilationRequirement
      \default Sensible

Comment (rraustad) [reply src/EnergyPlus/DataSizing.cc]: @mjwitte it would seem the same data would be used for all CoolingPeakLoadType inputs (to set ZoneCoolLoadSum and AvgZoneTemp)?

File: src/EnergyPlus/DataSizing.cc
Comment (amirroth) [inline src/EnergyPlus/DataSizing.cc:462]: Is `TotalCoolingLoad` not an `enum class`? It should be.

File: src/EnergyPlus/DataSizing.cc
Comment (amirroth) [inline src/EnergyPlus/DataSizing.cc]: `enum class`?

File: src/EnergyPlus/DataSizing.cc
Comment (amirroth) [inline src/EnergyPlus/DataSizing.cc]: `format`

File: src/EnergyPlus/DataSizing.cc
Comment (amirroth) [inline src/EnergyPlus/DataSizing.cc:502]: Probably outside the scope for this PR but this has never stopped me before so ... given RVO we may want to transition to returning multiple values by `return`ing a `struct` rather than using reference parameters. I'm not sure what the cutoff for number for members in a "`return struct`" would be but my intuition is that four or less is probably safe. The triple of `Temp`, `HumRat`, and `MassFlow` seems to be quite common. Anyhoo, I've said it.

File: src/EnergyPlus/DataSizing.cc
Comment (mjwitte) [reply src/EnergyPlus/DataSizing.cc]: Does the "Type of Load to Size On" get used for the system air flow rate, or the system coil sizing, or both?

Comment (rraustad) [reply src/EnergyPlus/DataSizing.cc]: Here it's being used to set the coil exit conditons used for sizing. I think this IF block was originally intended to do something different whether sensible or total but then that never panned out. In SimAirServingZones these are used to choose sensible or total load to set peak time and other variables. I think I'll just remove the IF block while fixing other suggestions.

Comment (mjwitte) [reply src/EnergyPlus/DataSizing.cc]: Hmmm, is there somewhere else that does something different for Total vs Sensible?

Comment (rraustad) [reply src/EnergyPlus/DataSizing.cc]: OK, I will look for code that does something different for total and sensible and check whether this code looks correct, needs updating, or can be removed completely.

Comment (rraustad) [reply src/EnergyPlus/DataSizing.cc]: SumZoneCoolLoadSeq is the sum of zone sensible cooling loads.

    // sum up the zone cooling load to be met by this system for this time step
    state.dataSize->SysSizing(state.dataSize->CurOverallSimDay, AirLoopNum).SumZoneCoolLoadSeq(TimeStepInDay) +=
        zoneSizing.CoolLoadSeq(TimeStepInDay);

And the calculation of DesExitTemp is a sensible calculation so it doesn't matter what the CoolingPeakLoadType is, the calculation would be the same. 

    DesExitTemp =
        max(finalSysSizing.CoolSupTemp, AvgZoneTemp - ZoneCoolLoadSum / (state.dataEnvrn->StdRhoAir * CpAir * finalSysSizing.DesCoolVolFlow));

File: src/EnergyPlus/DataSizing.cc
Comment (rraustad) [reply src/EnergyPlus/DataSizing.cc:462]: There is an issue here. The code is mixing these constexpr's. May take some time to decipher.

    // parameters for Cooling Peak Load TYpe
    constexpr int SensibleCoolingLoad(1);
    constexpr int TotalCoolingLoad(2);

    // parameters for Type of Load to Size On
    constexpr int Sensible(0);
    constexpr int Latent(1);
    constexpr int Total(2);
    constexpr int Ventilation(3);

E.g.:

    if (FinalSysSizing(iAirLoop).CoolingPeakLoadType == DataSizing::SensibleCoolingLoad) {
    } else if (FinalSysSizing(iAirLoop).CoolingPeakLoadType == DataSizing::Ventilation) {

File: src/EnergyPlus/DataSizing.cc
Comment (mjwitte) [reply src/EnergyPlus/DataSizing.cc:459]: Good catch.

File: src/EnergyPlus/DataSizing.cc
Comment (mjwitte) [reply src/EnergyPlus/DataSizing.cc:462]: Looks like this enum cleanup will wait for another PR.

File: tst/EnergyPlus/unit/DataSizing.unit.cc
Comment (mjwitte) [inline tst/EnergyPlus/unit/DataSizing.unit.cc]: Since you need to touch the code anyway, desing--> design.

File: tst/EnergyPlus/unit/DataSizing.unit.cc
Comment (mjwitte) [inline tst/EnergyPlus/unit/DataSizing.unit.cc:425]: Comprehensive unit test.

**Review by mjwitte — APPROVED** [review]
Overall, this looks good.
Defect files now have the same coil sizing for cooling coil GUID12 Sys20SCC-1.
There is an unused variable build warning.
Once that is fixed, this can merge.

**Comment by rraustad** [comment]
@mjwitte I don't see a build warning. Pulled in develop and corrected spelling per comment. I will commit after unit tests complete.

**Comment by mjwitte** [comment]
@rraustad This was the build warning [back on Nov 1](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-ab7027f6e3cabc077051d99a960e6b020e344918-x86_64-Linux-Ubuntu-18.04-gcc-7.5.html)
```
src/EnergyPlus/DataSizing.cc:455:9 | unused variable ‘PeakLoadType’ [-Wunused-variable]
```

**Comment by mjwitte** [comment]
Oh, never mind, that's from a different branch but showed up here on this page. Never mind, will wait for the updated push and let CI fire away.

**Comment by rraustad** [comment]
@Myoldmopar @mjwitte Only failure on Linux for SolarShadingTestGPU. The rest of CI is clean. This is ready to merge.

**Comment by Myoldmopar** [comment]
Thanks @rraustad, will do local testing and final review either later tonight or in the morning.

File: src/EnergyPlus/DataSizing.hh
Comment (Myoldmopar) [inline src/EnergyPlus/DataSizing.hh:117]: :+1:

File: tst/EnergyPlus/unit/UnitarySystem.unit.cc
Comment (Myoldmopar) [reply tst/EnergyPlus/unit/UnitarySystem.unit.cc:278]: Got it.

**Review by Myoldmopar — APPROVED** [review]

**Comment by Myoldmopar** [comment]
Changes look good, thanks for the bit of refactoring you added into it, as every little bit helps.  Local testing looks good, and given the nature of the issue and the changes, it's no surprise our CI is all clean.  I tried running the defect files, beginning with diff-ing the baseline and run files to see the order change, but unfortunately, all the unique identifiers, the files are completely different.  Anyway, this seems good to go in.  Merging.  Thanks @rraustad

**Pull request merged**
